include/irr/asset/filters/CSummedAreaTableImageFilter.h

```diff
@@ -0,0 +1,398 @@
+// Copyright (C) 2020 - AnastaZIuk
+// This file is part of the "IrrlichtBAW" engine.
+// For conditions of distribution and use, see copyright notice in irrlicht.h
+
+#ifndef __IRR_C_SUMMED_AREA_TABLE_IMAGE_FILTER_H_INCLUDED__
+#define __IRR_C_SUMMED_AREA_TABLE_IMAGE_FILTER_H_INCLUDED__
+
+#include "irr/core/core.h"
+
+#include <type_traits>
+#include <functional>
+
+#include "irr/asset/filters/CMatchedSizeInOutImageFilterCommon.h"
+#include "CConvertFormatImageFilter.h"
+
+namespace irr
+{
+namespace asset
+{
+
+template<bool ExclusiveMode>
+class CSummedAreaTableImageFilterBase
+{
+	public:
+		class CSummStateBase
+		{
+			public:
+
+				static inline constexpr size_t decodeTypeByteSize = sizeof(double);
+				uint8_t*	scratchMemory = nullptr;										//!< memory covering all regions used for temporary filling within computation of sum values
+				size_t	scratchMemoryByteSize = {};											//!< required byte size for entire scratch memory
+				bool normalizeImageByTotalSATValues = false;								//!< after sum performation division will be performed for the entire image by the max sum values in (maxX, 0, z) depending on input image - needed for UNORM and SNORM
+				
+				static inline size_t getRequiredScratchByteSize(const ICPUImage* inputImage, const ICPUImage* outputImage)
+				{
+					const auto& inputCreationParams = inputImage->getCreationParameters();
+					const auto channels = asset::getFormatChannelCount(inputCreationParams.format);
+					const auto& trueExtent = inputCreationParams.extent;
+
+					size_t retval = cachesOffset = trueExtent.width * trueExtent.height * trueExtent.depth * channels * decodeTypeByteSize;
```

Comment:
why does it require scratch size to be fit whole input image's texels? You need only just enough space to fit texels of processed sub-image (`state->extent`, not `inImage->extent`)

Reply:
I think I've repeated it 5x times already

Reply:
corrected in following commit